File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        let res = []
        candidates.sort((a,b)=>a-b);
        function dfs(curr, sum, i){
            if(sum === target){
                res.push([...curr]);
                return
            }
            if(i>=candidates.length || sum > target){
                return
            }
            dfs([...curr, candidates[i]], sum + candidates[i], i+1)
            while( i< (candidates.length-1) && candidates[i+1]==candidates[i]){
                i++;
            }
            dfs([...curr], sum , i+1)

        }
        dfs([], 0, 0)
        return res
    }
}

File: Data Structures & Algorithms/decode-ways/submission-1.py
class Solution:
    def numDecodings(self, s: str) -> int:
        cache = { len(s): 1 }
        def dfs(curr_idx):
            if curr_idx in cache:
                return cache[curr_idx]
            if curr_idx > len(s) or s[curr_idx] == "0":
                return 0
            
            res = dfs(curr_idx+ 1)
            if (curr_idx +1) < len(s) and int(s[curr_idx] + s[curr_idx+1]) <= 26:
                res += dfs(curr_idx + 2)
            cache[curr_idx] = res
            return res

        return dfs(0)
        